compiler/src/iree/compiler/Codegen/Common/TopkLowering.cpp

```diff
@@ -0,0 +1,915 @@
+                  ValueRange{idx0, ifExpand.getResult(0)});
+              addElemsElse =
+                  b.create<arith::AddIOp>(loc, ifExpand.getResult(0), one);
+              outValsElse = shiftElemsLoop.getResult(0);
+              outIndsElse = shiftElemsLoop.getResult(1);
+
+              // Insert the element in the output.
+              outValsElse = b.create<tensor::InsertOp>(
+                  loc, elemToInsertIfNeeded, outValsElse,
+                  ValueRange{idx0, insertionIndex});
+              outIndsElse = b.create<tensor::InsertOp>(
+                  loc,
+                  b.create<arith::IndexCastOp>(loc, b.getIntegerType(32),
+                                               elemIndex),
+                  outIndsElse, ValueRange{idx0, insertionIndex});
+              SmallVector<Value> newOperandsElse;
+              newOperandsElse.push_back(outValsElse);
+              newOperandsElse.push_back(outIndsElse);
+              newOperandsElse.push_back(smElemElse);
+              newOperandsElse.push_back(addElemsElse);
+              b.create<scf::YieldOp>(loc, newOperandsElse);
+            });
+
+        // Propagate returns from nested loops and ifs to the
+        // output of the function.
+        Value outValsAtEnd = ifAddAtEnd.getResult(0);
+        Value outIndsAtEnd = ifAddAtEnd.getResult(1);
+        Value smElemAtEnd = ifAddAtEnd.getResult(2);
+        Value addElemsAtEnd = ifAddAtEnd.getResult(3);
+
+        SmallVector<Value> newOperands;
+        newOperands.push_back(outValsAtEnd);
+        newOperands.push_back(outIndsAtEnd);
+        newOperands.push_back(smElemAtEnd);
+        newOperands.push_back(addElemsAtEnd);
+        b.create<scf::YieldOp>(loc, newOperands);
+      });
+  // Build and return the continuation of the outer loop,
+  // after inserting the element.
+  return InsertElemContinuationValues{
+      ifAddElem.getResult(0), ifAddElem.getResult(1), ifAddElem.getResult(2),
+      ifAddElem.getResult(3)};
+}
+
+/// Vectorize a `topKOp` with (1) static result and input types
+static LogicalResult lowerAsLinalgExtTopkUsingSCF(
+    RewriterBase &rewriter, IREE::LinalgExt::TopkOp topkOp,
+    ArrayRef<int64_t> inputVectorSizes, SmallVectorImpl<Value> &newResults) {
+  OpBuilder::InsertionGuard g(rewriter);
+  rewriter.setInsertionPoint(topkOp);
+  Location loc = topkOp.getLoc();
+  bool expectedRanksAndDims = true;
+  ReifiedRankedShapedTypeDims reifiedReturnShapes;
+  LogicalResult status =
+      cast<ReifyRankedShapedTypeOpInterface>(topkOp.getOperation())
+          .reifyResultShapes(rewriter, reifiedReturnShapes);
+  assert(succeeded(status) && "failed to reify result shapes");
+
+  if (failed(status))
+    return failure();
+
+  auto out0 = topkOp.getResults()[0];
+  auto resShapedTy0 = llvm::cast<ShapedType>(out0.getType());
+  auto out1 = topkOp.getResults()[1];
+  auto resShapedTy1 = llvm::cast<ShapedType>(out1.getType());
+
+  scf::ForOp scfInputLoop =
```

Comment:
Why do you expect the `scf.for` op to exist before hand?

Reply:
@MaheshRavishankar Thanks for the chat! It was useful to async on thinks I thought...

Reply:
@MaheshRavishankar 
>>> Why do you expect the scf.for op to exist before hand?
This is the loop that the tiling produces for the TopK operartion... Typically is one tile for each row (vector) in the 2-D case and one tile for the vector case for the whole input. This loop is used as a CodeGen point to generate the helper flags for the SCF lowering of the topk.
The loop is already there - looping over the input elements.
If the loop is not found, the transformation fails and it is delegated to the current implementation that will signal an error and fail.